### PR TITLE
base PR for set polarion as automated

### PR DESCRIPTION
**Pull request by dbasunag**
Reverts RedHatQE/python-utility-scripts#34

### CODE REVIEW

**Comment by msi-b1** [comment]
Report bugs in [Issues](https://github.com/myakove/github-webhook-server/issues)

The following are automatically added:
 * Add reviewers from OWNER file (in the root of the repository) under reviewers section.
 * Set PR size label.
 * New issue is created for the PR. (Closed when PR is merged/closed)
 * Run [pre-commit](https://pre-commit.ci/) if `.pre-commit-config.yaml` exists in the repo.

Available user actions:
 * To mark PR as WIP comment `/wip` to the PR, To remove it from the PR comment `/wip cancel` to the PR.
 * To block merging of PR comment `/hold`, To un-block merging of PR comment `/hold cancel`.
 * To mark PR as verified comment `/verified` to the PR, to un-verify comment `/verified cancel` to the PR.
        verified label removed on each new commit push.
 * To cherry pick a merged PR comment `/cherry-pick <target branch to cherry-pick to>` in the PR.
    * Multiple target branches can be cherry-picked, separated by spaces. (`/cherry-pick branch1 branch2`)
    * Cherry-pick will be started when PR is merged
 * To build and push container image command `/build-and-push-container` in the PR (tag will be the PR number).
 * To add a label by comment use `/<label name>`, to remove, use `/<label name> cancel`
 * To assign reviewers based on OWNERS file use `/assign-reviewers`
 * To check if PR can be merged use `/check-can-merge`

<details>
<summary>Supported /retest check runs</summary>

 * `/retest tox`: Retest tox
 * `/retest python-module-install`: Retest python-module-install

</details>

<details>
<summary>Supported labels</summary>

 * hold
 * verified
 * wip
 * lgtm

</details>

File: apps/polarion/polarion_set_automated.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    if added_polarions := update_polarion_ids(
+        polarion_ids=added_ids, project_id=polarion_project_id, is_automated=True, is_approved=True
+    ):
+        LOGGER.debug(f"Following polarion ids were marked automated and approved: {added_polarions.get('updated')}")
+    if removed_polarions.get("failed") or added_polarions.get("failed"):
+        LOGGER.error(
+            f"Following polarion ids updates failed. Removed ids: {removed_polarions.get('failed')}"
+            f"Added ids:: {added_polarions.get('failed')}"
+        )
+        raise click.Abort()
```
Comment (rnetser) [inline apps/polarion/polarion_set_automated.py:49]: please convert `click.Abort()` to custom exceptions.

File: apps/polarion/polarion_set_automated.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    "--config-file-path",
+    help="Provide absolute path to the config file. Any CLI option(s) would override YAML file",
+    type=click.Path(),
+    default=os.path.expanduser("~/.config/python-utility-scripts/config.yaml"),
+)
+@click.option("--project-id", "-p", help="Provide the polarion project id")
+@click.option("--verbosity", default=False, is_flag=True)
+def polarion_approve_automate(config_file_path, project_id, verbosity):
+    if verbosity:
+        LOGGER.setLevel(logging.DEBUG)
```
Comment (rnetser) [inline apps/polarion/polarion_set_automated.py:23]:
```suggestion
    if not verbose:
        logging.disable(logging.CRITICAL)
```

and then you do not need debug

**Review by rnetser — CHANGES_REQUESTED** [review]

File: apps/polarion/polarion_utils.py [hunk 3]
```diff
@@ -19,31 +25,74 @@ def git_diff_lines():
...
+
+def find_polarion_ids(polarion_project_id, string_to_match):
+    return re.findall(
+        rf"pytest.mark.polarion.*({polarion_project_id}-[0-9]+)",
+        "\n".join(git_diff_lines().get(string_to_match, [])),
+        re.MULTILINE | re.IGNORECASE,
+    )
+
+
+def get_polarion_project_id(project_id, config_file_path, util_name):
```
[the remark this answers is not in the document]
Comment (dbasunag) [reply apps/polarion/polarion_utils.py:68]: Util name is required param. But the other two would be set to None as either of those can be passed.

File: apps/polarion/polarion_set_automated.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+@click.command()
+@click.option(
+    "--config-file-path",
+    help="Provide absolute path to the config file. Any CLI option(s) would override YAML file",
+    type=click.Path(),
+    default=os.path.expanduser("~/.config/python-utility-scripts/config.yaml"),
+)
+@click.option("--project-id", "-p", help="Provide the polarion project id")
+@click.option("--verbosity", default=False, is_flag=True)
+def polarion_approve_automate(config_file_path, project_id, verbosity):
```
[the remark this answers is not in the document]
Comment (dbasunag) [reply apps/polarion/polarion_set_automated.py:21]: Creating a separate PR for this one.

File: apps/polarion/polarion_set_automated.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    if verbosity:
```
[the remark this answers is not in the document]
Comment (dbasunag) [reply apps/polarion/polarion_set_automated.py]: I am not clear on this one. These won't need ipdb/pdb for debugging failures. If they fail, it is due to polarion connectivity or not being able to access/update polarions. So even with function_runner_with_pdb(), I would need to pass `--pdb` to default to false. I don't see a benefit of using it here. Please advice.

[the remark this answers is not in the document]
Comment (dbasunag) [reply apps/polarion/polarion_set_automated.py]: Kept this as per chat.

**Comment by dbasunag** [comment]
@myakove / @rnetser I had to change the coverage fail_under percentage. With polarion not being tested that well, current coverage  67.51% won't meet 75% threshold, may be in future it can be improved for other areas. Please let me know what you think.

File: apps/polarion/polarion_set_automated.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    type=click.Path(),
+    default=os.path.expanduser("~/.config/python-utility-scripts/config.yaml"),
+)
+@click.option("--project-id", "-p", help="Provide the polarion project id")
+@click.option("--verbose", default=False, is_flag=True)
+def polarion_approve_automate(config_file_path: str, project_id: str, verbose: bool) -> None:
+    if verbose:
+        LOGGER.setLevel(logging.DEBUG)
+    else:
+        logging.disable(logging.WARNING)
```
Comment (myakove) [inline apps/polarion/polarion_set_automated.py:50]: you should use logger info and not debug in all your logs and here you should enable logs if verbose else disable all logs.

File: apps/polarion/polarion_utils.py [hunk 7]
```diff
@@ -18,30 +26,75 @@ def git_diff_lines() -> Dict[str, List[str]]:
...
+        "\n".join(git_diff_lines().get(string_to_match, [])),
+        re.MULTILINE | re.IGNORECASE,
+    )
+
+
+def get_polarion_project_id(util_name: str, config_file_path: str) -> str:
+    polarion_project_id = get_util_config(util_name=util_name, config_file_path=config_file_path).get("project_id")
+    if not polarion_project_id:
+        LOGGER.error("Polarion project id must be passed via config file or command line")
+        raise click.Abort()
```
Comment (myakove) [inline apps/polarion/polarion_utils.py:73]: click.Abort() exception is bad exception, please create your own exception that can have msg.

**Review by myakove — CHANGES_REQUESTED** [review]

File: apps/polarion/polarion_set_automated.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+from apps.polarion.exceptions import PolarionTestCaseApprovalFailed
+from apps.polarion.polarion_utils import get_polarion_project_id, find_polarion_ids, update_polarion_ids
+from typing import List, Dict, Optional, Iterable
+
+
+LOGGER = get_logger(name=__name__)
+
+
+def approve_tests(polarion_project_id: str, added_ids: Iterable[str]) -> Dict[str, List[str]]:
```
[the remark this answers is not in the document]
Comment (dbasunag) [reply apps/polarion/polarion_set_automated.py:16]: mypy did not like it originally, so I had to change it to Iterable. But it has been a while, so I will retry and let you know.

[the remark this answers is not in the document]
Comment (dbasunag) [reply apps/polarion/polarion_set_automated.py:16]: Updated.

File: apps/polarion/polarion_utils.py (hunk 7 above)
Comment (dbasunag) [reply apps/polarion/polarion_utils.py:73]: Replaced with exceptions.

File: apps/polarion/polarion_set_automated.py (hunk 6 above)
Comment (dbasunag) [reply apps/polarion/polarion_set_automated.py]: Ok, please let me know if the changes I made is aligned to what you want.

Comment (dbasunag) [reply apps/polarion/polarion_set_automated.py]: I left `logging.disable(logging.ERROR)` as otherwise there would be no debugging on failure.

**Review by rnetser — APPROVED** [review]

**Comment by redhat-qe-bot1** [comment]
Pull request already merged, not running verified

**Pull request merged**